### ISSUE TITLE
Ungraceful exit with "Most likely this is a multi-threaded client and XInitThreads has not been called"

### ISSUE THREAD

**Issue by Dominaezzz** [JetBrains/compose-jb#749]
With this program.
```kotlin
@OptIn(ExperimentalComposeUiApi::class)
fun main() {
    application {
        val windowState = rememberWindowState()
        Window(windowState, title = "New Window API playground") {
            MenuBar {
                Menu("File") {
                    Item("Open", onClick = { println("Open!!!") })
                }

                Menu("Edit") {
                    Item("Copy", onClick = { println("Copy stuff!!!") })

                    var isOpen by remember { mutableStateOf(false) }
                    LaunchedEffect(Unit) {
                        while (isActive) {
                            delay(2000)
                            isOpen = !isOpen
                        }
                    }

                    if (isOpen) {
                        Item("Paste", onClick = { println("Paste stuff!!!") })
                    }
                }
            }
        }
    }
}
```

If you open the menu long enough for it to toggle, then close the window, you get this error message.
```
[xcb] Unknown sequence number while processing reply
[xcb] Most likely this is a multi-threaded client and XInitThreads has not been called
[xcb] Aborting, sorry about that.
java: xcb_io.c:725: _XReply: Assertion `!xcb_xlib_threads_sequence_lost' failed.
```

Compose Desktop: 0.4.0
Kotlin: 1.5.10
OS: Arch Linux (i3, X11)

**Comment by Dominaezzz** [comment]
Was working on a different application, without menu and it consistently closes like this now.

**Comment by igordmn** [comment]
Does the crash reproduce with an empty windows?

old API:
```
import androidx.compose.desktop.Window

fun main() {
    Window {

    }
    Window {

    }
}
```

new API:
```
import androidx.compose.ui.ExperimentalComposeUiApi
import androidx.compose.ui.window.Window
import androidx.compose.ui.window.application

@OptIn(ExperimentalComposeUiApi::class)
fun main() = application {
    Window {

    }
    Window {

    }
}
```

**Comment by Dominaezzz** [comment]
The old API closes just fine.
The new API crashes 90% of the time.

Are you not able to reproduce this on your end?

**Comment by igordmn** [comment]
I reproduced this with menu, but can't reproduce this without menu (Ubuntu 20.04 X11, OpenJDK 15). The snippets above work without a crash.

Not sure, but probably in the old API it doesn't crash because we call System.exit after the last window close before application properly dispose resources.

**Comment by eygraber** [comment]
I get this without the menu on Ubuntu 20.04 (using 0.5.0-build270). It just started happening randomly, without any code changes. Until now it ran without that error when it exits. JDK 11.

**Comment by Omnieboer** [comment]
I'm getting this on a (pretty) fresh install of Linux Mint 20.2 Cinnamon.
I have never used a Menu, but I am multi-threading (or at least, coroutining?) because I'm running a UI and TCP client at the same time. Should I actually try to call `XInitThreads` or await a bug fix?

**Comment by igordmn** [comment]
> Should I actually try to call XInitThreads or await a bug fix?

Hard to tell without a reproducer. I haven't able to reproduce this without Menu. But the crash with Menu is probably a different thing.

**Comment by Dominaezzz** [comment]
Updated the issue title with the main bit. Can be more specific when we know where the issue is.

Compose Desktop isn't multi threaded, at least the bit that handles swing stuff.
I reckon some threading was accidently introduced somewhere but it's hard to tell. I've looked at the code myself and it seems mostly fine.
What I find most interesting is that this happens on exit only. Which is an area I haven't actually looked at.

**Comment by Omnieboer** [comment]
It seems my TCP thread exits gracefully, but then the UI thread gives this problem. But I'm not sure at which point it actually happens.
```kotlin
println("Closing UI")
applicationScope.exitApplication()
println("Closed all")
```
still prints "Closed all"

**Comment by olonho** [comment]
Maybe `exitProcess` shall be called on closing the last window.

**Comment by igordmn** [comment]
Should be fixed in 1.0.0-beta6-dev455.

Now we call `exitProcess` in the end of the `application`, which will bypass the issue in JDK (or GTK, not sure)

**Comment by okushnikov** [comment]
Please check the <a href="https://youtrack.jetbrains.com/issue/CMP-6230">following ticket</a> on YouTrack for follow-ups to this issue. GitHub issues will be closed in the coming weeks.
